tests/test_timer.cpp

```diff
@@ -32,164 +33,555 @@ using namespace testing;
 };
 } // namespace
 
-TEST_F(Timer, times_out)
+TEST_F(TestTimer, times_out)
 {
-    mpu::Timer t{duration_cast<seconds>(1ms), [this]() {
+    EXPECT_CALL(*mock_timer_sync_funcs, wait_for(_, _, _)).WillOnce([this](auto&, auto& timer_lock, auto&) {
+        std::unique_lock<std::mutex> test_lock(cv_m);
+
+        // Unlock the Timer's internal lock
+        timer_lock.unlock();
+
+        // Wait for notification from main test thread
+        cv.wait(test_lock, [this] { return test_notify_called.load(); });
+
+        // Re-lock the Timer's internal lock before returning
+        timer_lock.lock();
+
+        return std::cv_status::timeout;
+    });
+
+    // notify_all is called twice via the stop() in start() and the stop() in the dtor
+    EXPECT_CALL(*mock_timer_sync_funcs, notify_all(_)).Times(2).WillRepeatedly(Return());
+
+    mpu::Timer t{default_timeout, [this]() {
                      {
-                         std::lock_guard<std::mutex> lk{cv_m};
+                         std::lock_guard<std::mutex> test_lock{cv_m};
                          timedout.store(true);
                      }
                      cv.notify_all();
                  }};
-    std::unique_lock<std::mutex> lk(cv_m);
 
     t.start();
-    ASSERT_FALSE(timedout.load()) << "Should not have timed out yet";
 
-    cv.wait_for(lk, 10ms); // Windows CI needs longer...
+    std::unique_lock<std::mutex> test_lock(cv_m);
+    test_notify_called.store(true);
 
-    ASSERT_TRUE(timedout.load()) << "Should have timed out";
+    // Notify the mocked wait_for()
+    cv.notify_all();
+
+    ASSERT_TRUE(cv.wait_for(test_lock, default_wait, [this] { return timedout.load(); }))
+        << "Did not time out in reasonable timeframe";
 }
 
-TEST_F(Timer, stops)
+TEST_F(TestTimer, stops)
 {
-    mpu::Timer t{duration_cast<seconds>(10s), [this]() { timedout.store(true); }};
+    EXPECT_CALL(*mock_timer_sync_funcs, wait_for(_, _, _)).WillOnce([this](auto&, auto& timer_lock, auto&) {
+        std::unique_lock<std::mutex> test_lock(cv_m);
+
+        // Unlock the Timer's internal lock
+        timer_lock.unlock();
+
+        // Notify the test thread that the timer is running
+        timer_running.store(true);
+        cv.notify_all();
+
+        // Wait to be notified by the mock notify_all()
+        cv.wait(test_lock, [this] { return test_notify_called.load(); });
+
+        // Re-lock the Timer's internal lock before returning
+        timer_lock.lock();
+
+        return std::cv_status::no_timeout;
+    });
+
+    EXPECT_CALL(*mock_timer_sync_funcs, notify_all(_))
+        .WillOnce(Return()) // Handles the stop() in start()
+        .WillOnce(          // Handles the stop() called in the test
+            [this](auto&) {
+                {
+                    std::unique_lock<std::mutex> test_lock(cv_m);
+                    test_notify_called.store(true);
+                }
+                // Notify the mocked wait_for()
+                cv.notify_all();
```

Comment:
Ok, sure, will look into that.

Reply:
I couldn't use a `ONCALL().WillByDefault()` because two separate actions are needed per test, but I did make a general `EXPECT_CALL().WillOnce().WillRepeatedly()` for this.  The unfortunate side effect is that I can't test the number of times it is called.  Let me know what you think.

Reply:
I see, thanks. I am not sure if you could just `EXPECT_CALL` again, adding only `Times`, but that would be kind of redundant. There is an `Expectation` handle that can be used to hold on to expectations, but this appears to be useful only for `After` clauses (no interface to ask `Times`). 

You might be able to turn `test_notify_called` into `test_notify_count`, which tests could then verify. The var is used for other purposes so maybe another one would be needed. And something else would be needed for the dtor call. Or you could extract into a helper method receiving the cardinality and call it in tests.

I think it's ok as it is, but up to you.

Reply:
Yeah, I thought of the `EXPECT_CALL` again with only adding `Times` and that overrides the other `EXPECT_CALL`'s and uses the default return value, so long story short, it breaks things :wink: 

I think it's fine the way it is too since it's not *really* important to test how many times it is called.